### PR TITLE
Css 10800/local charm

### PR DESCRIPTION
**Pull request by SimoneDutto**
## Description

This PR adds the endpoints to deploy local charms. They are HTTP endpoints and need to be proxyed as they are to the right controller.

Fixes _JIRA/GitHub issue number_

## Engineering checklist
*Check only items that apply*

- [ ] Documentation updated
- [ ] Covered by unit tests
- [ ] Covered by integration tests

### CODE REVIEW

File: cmd/jimmsrv/service/register.go
Comment (pkulik0) [inline cmd/jimmsrv/service/register.go]: Just realized jimm still uses the regex for goheader which is buggy. I'll make a PR to use the same built-in value as in the LP repos but you'll need to look out for this before v3 is merged into this feature branch.

File: internal/jimmhttp/http.go
Comment (pkulik0) [inline internal/jimmhttp/http.go:28]: new line

File: cmd/jimmsrv/service/register.go
Comment (pkulik0) [inline cmd/jimmsrv/service/register.go]: nit: we could reuse the same one on line 23 & 27 but it's cheap to create so it's ok

File: internal/jimmhttp/http_test.go
Comment (pkulik0) [inline internal/jimmhttp/http_test.go]: we could just call `http.Get(url)` without the instantiation of http.Client above

File: internal/jimmtest/mocks/model.go
Comment (pkulik0) [inline internal/jimmtest/mocks/model.go]: should check j.GetModel_ for nil instead

File: internal/jujuapi/httpproxy_integration_test.go
Comment (pkulik0) [inline internal/jujuapi/httpproxy_integration_test.go:207]: new line at the end

**Review by pkulik0 — COMMENTED** [review]
nice work, only some minor comments and missing new lines

File: internal/jujuapi/httpproxy.go
Comment (pkulik0) [inline internal/jujuapi/httpproxy.go:49]: new line at the end too, somehow not all my comments saved

File: internal/jimmhttp/http_test.go
Comment (pkulik0) [inline internal/jimmhttp/http_test.go:45]: \n

**Comment by SimoneDutto** [comment]
@pkulik0 for the new line, strangely enough, I have this line in my files, but the new line is not there in the commit.
Do I need to change some config?

File: cmd/jimmsrv/service/register.go
Comment (babakks) [inline cmd/jimmsrv/service/register.go]: Do all endpoints require the `writer` permission?

File: internal/jimmhttp/http.go
Comment (babakks) [inline internal/jimmhttp/http.go]: It's a bit nicer to put this at the top of the file.

File: internal/jujuapi/httpproxy.go
Comment (babakks) [inline internal/jujuapi/httpproxy.go]: Can't we delete the comment?

File: internal/jujuapi/httpproxy.go
Comment (babakks) [inline internal/jujuapi/httpproxy.go]: Unnecessary comment.

File: internal/jujuapi/httpproxy.go
Comment (babakks) [inline internal/jujuapi/httpproxy.go]: I think a good comment here would be to explain why we need the credentials (which is because the HTTP endpoints require credentials passed as Basic auth).

File: internal/jujuapi/httpproxy_integration_test.go
Comment (babakks) [inline internal/jujuapi/httpproxy_integration_test.go]: It's better to make `setup` field, optional, and in the loop you can check if `test.setup` is not `nil`, and then call it.

File: internal/jujuapi/httpproxy_integration_test.go
Comment (babakks) [inline internal/jujuapi/httpproxy_integration_test.go]: Can you please add a `expectedBody` field to assert it against the returned HTTP response body?

File: internal/jujuapi/httpproxy_integration_test.go
Comment (babakks) [inline internal/jujuapi/httpproxy_integration_test.go]: And here `expectedBody` should be `"OK"`.

File: internal/jujuapi/httpproxy_integration_test.go
Comment (babakks) [inline internal/jujuapi/httpproxy_integration_test.go]: If possible, please don't use the word "dummy", because it's against the inclusive language.

File: internal/jujuapi/httpproxy_integration_test.go
Comment (babakks) [inline internal/jujuapi/httpproxy_integration_test.go]: And here.

**Review by babakks — APPROVED** [review]
Great job! Thanks for this. Just one note, in JIMM we don't separate unit and integration tests by filename. So, you can just rename the newly added file to `httpproxy_test.go`.

File: internal/jimmhttp/http_test.go
Comment (babakks) [inline internal/jimmhttp/http_test.go]: I think you don't need to call `resp.Body.Close()` if you don't look at the `Body` field. From the `Response.Body` docs:

> The response body is streamed on demand as the Body field is read. If the network connection fails or the server terminates the response, Body.Read calls return an error.

**Review by babakks — DISMISSED** [review]

File: cmd/jimmsrv/service/register.go
Comment (ale8k) [inline cmd/jimmsrv/service/register.go]: One below breaks return to newline, lets keep it consistent

File: cmd/jimmsrv/service/register.go
Comment (ale8k) [inline cmd/jimmsrv/service/register.go]: No need for this newline

File: internal/jimmhttp/http.go
Comment (ale8k) [inline internal/jimmhttp/http.go]: We have a struct called HTTPProxier in httpproxy.go of jujuapi, naming is confusing here

File: cmd/jimmsrv/service/register.go
Comment (ale8k) [inline cmd/jimmsrv/service/register.go]: Why are we embedding the proxier in http handler? It already has a ServeHTTP but signature is different, why not just make the proxier an actual handler instead of embedding it?...

File: internal/jimmhttp/http.go
Comment (ale8k) [inline internal/jimmhttp/http.go]: We have a servehttp here with the wrong signature and embed it in the handler, why?

File: internal/jimmhttp/http.go
Comment (ale8k) [inline internal/jimmhttp/http.go]: This is called HTTPHandler but can only proxy according to its servehttp? This is very confusing?

File: cmd/jimmsrv/service/service.go
Comment (ale8k) [inline cmd/jimmsrv/service/service.go]: Shouldn't we make methods for:

- RegisterModelWSEndpoint
- RegisterControllerWSEndpoint
- RegisterReBACAdminEndpoints
- RegisterDebugEndpoint
- RegisterMetricsEndpoint

But still, I think having the group top level and explaining what middlewares it has so we can move routes more easily around depending on the middleware they need is better. I.e., all need tracing, and logging, so a top level group with those middle wares would embed other groups. Let's revisit this together

File: internal/jujuapi/httpproxy.go
Comment (ale8k) [inline internal/jujuapi/httpproxy.go]: Interface and struct named the same, this is confusing... Let's refactor some of this together

File: internal/jujuapi/httpproxy.go
Comment (ale8k) [inline internal/jujuapi/httpproxy.go]: This will be a consistent error message we can use, let's not make an inline function for something that can be obviously used by any handler

File: internal/jujuapi/httpproxy.go
Comment (ale8k) [inline internal/jujuapi/httpproxy.go]: Thinking about this, the ProxyHTTP responds to the request on line 65, but this isn't right, onyl the handler should respond to requests based on errors returned. ProxyHTTP probably should return an error right?

**Review by ale8k — CHANGES_REQUESTED** [review]

**Review by ale8k — CHANGES_REQUESTED** [review]
Structuring is confusing, happy to refactor a bit together then I think it'll be ok. Also maybe lets change proxyhttp in RPC, don't think it necessarily belongs there as it has nothing to do with RPC thinking about it.

**Comment by pkulik0** [comment]
@SimoneDutto I did a small test and realized I was just wrong about the new lines. GitHub shows a red sign next to the line numbers when a new line is missing and when it's there they just omit it from the view. So turns out it was correct from the beginning.

File: cmd/jimmsrv/service/register.go
Comment (SimoneDutto) [reply cmd/jimmsrv/service/register.go]: yes, they do!

File: internal/jimmhttp/http_test.go
Comment (SimoneDutto) [reply internal/jimmhttp/http_test.go]: the lint complains about it, even if I don't use it

File: internal/jujuapi/httpproxy_integration_test.go
Comment (SimoneDutto) [reply internal/jujuapi/httpproxy_integration_test.go]: it's the default local charm in juju tests. If I use another name it doesn't work because it is not found.

[search in juju code](https://github.com/search?q=repo%3Ajuju%2Fjuju+testcharms.Repo.CharmArchive%28c.MkDir%28%29%2C+%22dummy%22%29&type=code)

File: internal/jujuapi/httpproxy_integration_test.go
Comment (SimoneDutto) [reply internal/jujuapi/httpproxy_integration_test.go]: same comment as above

File: internal/jimmhttp/package_test.go
Comment (kian99) [inline internal/jimmhttp/package_test.go:11]: Just a quick review, but I would like to see Go Check tests limited to the `jujuapi` package.

File: cmd/jimmsrv/service/service.go
Comment (babakks) [inline cmd/jimmsrv/service/service.go:456]: Nice!

File: internal/jimmhttp/httpproxy_handler.go
Comment (babakks) [inline internal/jimmhttp/httpproxy_handler.go:20]: Needs godoc.

File: internal/jimmhttp/httpproxy_handler.go
Comment (babakks) [inline internal/jimmhttp/httpproxy_handler.go:27]: Although it's somewhat clear, but would rather to see a godoc above it.

File: internal/jimmhttp/httpproxy_handler.go
Comment (babakks) [inline internal/jimmhttp/httpproxy_handler.go]: s/hhtp/http

File: internal/jimmhttp/httpproxy_handler.go
Comment (babakks) [inline internal/jimmhttp/httpproxy_handler.go]: Needs godoc.

File: internal/jimmhttp/package_test.go
Comment (SimoneDutto) [reply internal/jimmhttp/package_test.go:11]: my tests depends on jimmsuite which is from gocheck. Otherwise I agree

File: internal/jimmhttp/httpproxy_handler.go
Comment (babakks) [inline internal/jimmhttp/httpproxy_handler.go]: What if it's not a gateway timeout?

File: internal/jimmhttp/package_test.go
Comment (babakks) [reply internal/jimmhttp/package_test.go:11]: Yeah, it's needed to run `JIMMSuite`-based tests (the `httpProxySuite`).

File: internal/middleware/authn.go
Comment (babakks) [inline internal/middleware/authn.go:22]: Let's avoid stripped names in types/comments.

File: internal/jimmhttp/httpproxy_handler.go
Comment (SimoneDutto) [reply internal/jimmhttp/httpproxy_handler.go]: this method is not used, i'll remove it

File: internal/middleware/authn.go
Comment (SimoneDutto) [reply internal/middleware/authn.go:22]: the idea for this proxy function is to:
- no err -> return nil
- err -> log it and try with the next one
- if you reach the end it means no controller responded you so "GATEWAY TIMEOUT"

File: internal/jimmhttp/httpproxy_handler.go
Comment (SimoneDutto) [reply internal/jimmhttp/httpproxy_handler.go:77]: see https://github.com/canonical/jimm/pull/1375#discussion_r1774976420

Comment (SimoneDutto) [reply internal/jimmhttp/httpproxy_handler.go]: the idea for this proxy function is to:

    no err -> return nil
    err -> log it and try with the next one
    if you reach the end it means no controller responded you so "GATEWAY TIMEOUT"

I added this description to the godoc of the proxy method

File: cmd/jimmsrv/service/service.go
Comment (kian99) [reply cmd/jimmsrv/service/service.go:456]: What does this syntax do? New to me

File: internal/jimmhttp/httpproxy_handler.go
Comment (kian99) [inline internal/jimmhttp/httpproxy_handler.go:43]: Should this be exported?

File: internal/jimmhttp/httpproxy_handler.go
Comment (kian99) [inline internal/jimmhttp/httpproxy_handler.go:20]: Should this be exported? Probably not I think

**Review by kian99 — APPROVED** [review]
LGTM with some tweaks

File: internal/jimmhttp/httpproxy_handler.go
Comment (SimoneDutto) [reply internal/jimmhttp/httpproxy_handler.go:43]: It is to comply with HTTPHandler interface.
https://github.com/canonical/jimm/blob/85c03a97fbe54830e306706cca4174bff1f379bc/internal/jimmhttp/handler.go#L8-L12

File: cmd/jimmsrv/service/service.go
Comment (SimoneDutto) [reply cmd/jimmsrv/service/service.go:456]: this handler is mounter for endpoints formatted like:
- /model/{uuid}/charms
-  /model/{uuid}/applications

Nothing very fancy, just the way of doing or in `chi`

File: internal/jimmhttp/httpproxy_handler.go
Comment (kian99) [reply internal/jimmhttp/httpproxy_handler.go:43]: I think we should edit the interface. None of the structs that implement it, call it outside of the Routes() method. Which makes sense, If a struct has some required middleware, that is internal to the implementation of the handler.

Additionally, one could wrap the handler in more middleware.

File: internal/jimmhttp/httpproxy_handler.go
Comment (ale8k) [inline internal/jimmhttp/httpproxy_handler.go:26]: Comment should be more like // ProxyEndpoints sets the endpoints to be all on this handler.

File: internal/jimmhttp/httpproxy_handler.go
Comment (SimoneDutto) [reply internal/jimmhttp/httpproxy_handler.go:43]: I think both interface, struct and methods have been left to public to leave the consumer the ability of composing and calling the methods when he wants.

But @ale8k can answer to this better than me

File: internal/jimmhttp/httpproxy_handler_test.go
Comment (ale8k) [inline internal/jimmhttp/httpproxy_handler_test.go:69]: Would love space between lines called its so hard to read ...

File: internal/jimmhttp/httpproxy_handler_test.go
Comment (ale8k) [inline internal/jimmhttp/httpproxy_handler_test.go:64]: Why don't we mock jimm & credential store here? We don't need to hit the actual db no?

File: internal/middleware/authn.go
Comment (ale8k) [inline internal/middleware/authn.go:22]: Lets make a better name haha, @kian99 help

**Review by ale8k — APPROVED** [review]
nice! i have one minor concern around testing of http handlers ... we should mock everywhere we can really and separate integration tests... might need to explore a little on best way to test - lets discuss standup. I'm approving for now but plz don't merge until we discuss this topic

**Comment by ale8k** [comment]
Ignore the testing thing for now, ok to merge regardless but lets discuss it somepoint as a team

File: internal/jimmhttp/httpproxy_handler.go
Comment (SimoneDutto) [reply internal/jimmhttp/httpproxy_handler.go:43]: addressing this in a followup

**Pull request merged**
